### PR TITLE
2139 select dot star syntax

### LINKED ISSUE #2139: Integrate the "select-clause" dot star syntax
**Description**

Integrate the "select-clause" dot star syntax

The purpose of this issue, is to integrate into the parser, the "select-clause" dot star syntax described like that, using ANTLR notation

SELECT (ALL|DISTINCT) ? (Identifier (DOT Identifier)*) DOT_STAR
with
DOT = .
DOT_STAR = .*

_This issue must be implemented after issue #2137_

The integration must be similar like that:

-     Adapt the Scanner to recognize SQL Tokens of the syntax (DOT , DOT_STAR, ...)

-     Add a ANTLR CodeElement syntax: for the QualifiedIdentifier = (Identifier (DOT Identifier)*);

-     Add a CUP Node syntax:

	subSelect ::=  SelectClause selectList;

	selectList := selectListItem 
	| selectList Comma selectListItem;

	selectListItem := starExpression;

	starExpression ::= QualifiedIdentifier DotStar;

-    Mix #2137 CUP syntax rules with syntax rules of this issue.

### PR DESCRIPTION
**Pull request by laabidiHend**
Fixes #2139

### CODE REVIEW

File: TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs [hunk 1]
```diff
@@ -49,19 +52,73 @@ private SelectClause CreateSelectClause(CodeElementsParser.Sql_selectClauseConte
...
+                        else
+                        {
+                            QualifiedSymbolReference fullName = new QualifiedSymbolReference(SchemaName, tail);
+                            return new DotStarSelection(fullName);
+                        }
+                    }
+                    else
+                    {
+                        return new DotStarSelection(tail);
+                    }
```
Comment (fm-117) [inline TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs:116]: I think we should refactor this as follows:
- extract all three parts of the name into Token instances, the first 2 may be null
- call a method to create the appropriate SymbolReference given the three tokens. This new method will be reusable for column names for example

As for the `SymbolType`, this is not corect. `AlphabetName` is a Cobol concept and has no meaning in SQL. We have to create new symbol types. So add a `SqlTableOrViewOrCorrelationName` member in the enum and use it.

File: TypeCobol/Compiler/Sql/Scanner/SqlScanner.cs [hunk 2]
```diff
@@ -43,6 +43,12 @@ public override Token GetTokenStartingFrom(int startIndex)
                 case '*':
                     currentIndex++;
                     return new Token(TokenType.MultiplyOperator, startIndex, currentIndex - 1, tokensLine);
+                case ',':
+                    currentIndex++;
+                    return new Token(TokenType.SQL_CommaSeparator, startIndex, currentIndex - 1, tokensLine);
+                case '.':
+                    currentIndex++;
+                    return new Token(TokenType.PeriodSeparator, startIndex, currentIndex - 1, tokensLine);
```
Comment (fm-117) [inline TypeCobol/Compiler/Sql/Scanner/SqlScanner.cs:51]: For the tokens made of one char, you can use `ScanOneChar` method from Scanner. So move the method from `Scanner` to `AbstractScanner`, make it `protected` and then call it here for the three cases '*', ',' and '.'.

**Review by fm-117 — CHANGES_REQUESTED** [review]

File: TypeCobol/AntlrGrammar/CobolWords.g4 [hunk 3]
```diff
@@ -879,7 +879,8 @@ tokens
 	SQL_XMLCAST,
 	SQL_YEAR,
 	SQL_YEARS,
-	SQL_ZONE
+	SQL_ZONE,
+    SQL_CommaSeparator
```
Comment (smedilol) [inline TypeCobol/AntlrGrammar/CobolWords.g4:883]: Tokens are all in uppercase.
```suggestion
    SQL_COMMA_SEPARATOR
```
Can you change it, please?

File: TypeCobol.Test/Parser/CodeElements/EXEC.txt [hunk 4]
```diff
@@ -3,47 +3,69 @@
 
 [[ExecStatementText]] [3,8:UPDATE]<ExecStatementText> --> [3,8:UPDATE]<ExecStatementText>
 
+[[ExecStatementText]] [9,9: ]<ExecStatementText> --> [9,9: ]<ExecStatementText>
```
Comment (smedilol) [inline TypeCobol.Test/Parser/CodeElements/EXEC.txt:6]: Whitespace token where removed in last PR #2164 (commit b8da4855039ec51a4bb386fea075554afe254a82), why put them back in this PR ?

File: TypeCobol.Test/TestCollection.cs [hunk 5]
```diff
@@ -160,8 +160,8 @@ private void CheckParserCobol85(bool cobol)
             foreach (string directory in Directory.GetDirectories(root))
             {
                 var dirname = Path.GetFileName(directory);
-
-			    Console.WriteLine("Entering directory \"" + dirname + "\" [" + string.Join(", ", extensions) + "]:");
+                int number = Path.GetFileNameWithoutExtension(directory).Count();
```
Comment (smedilol) [inline TypeCobol.Test/TestCollection.cs:163]: Test code to be removed ?

**Review by smedilol — CHANGES_REQUESTED** [review]

File: TypeCobol/AntlrGrammar/CobolWords.g4 (hunk 3 above)
Comment (smedilol) [reply TypeCobol/AntlrGrammar/CobolWords.g4:883]: My bad, `CommaSeparator` is not in uppercase. It must be kept like that.

File: TypeCobol.Test/Parser/Programs/Cobol85/ExecSql/ExecSqlWithSelectStar.rdzMix.txt [hunk 6]
```diff
@@ -0,0 +1,17 @@
+﻿       IDENTIFICATION DIVISION.
+       PROGRAM-ID. DVZZMFT3.
+Line 3[8,21] <27, Error, Syntax> - Syntax error : extraneous input 'DATADIVISION.' expecting {ExecStatementText, ExecStatementEnd, CommitStatement, SelectStatement}
```
Comment (fm-117) [inline TypeCobol.Test/Parser/Programs/Cobol85/ExecSql/ExecSqlWithSelectStar.rdzMix.txt:3]: This error is abnormal. DATA DIVISION is correctly placed.

File: TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs [hunk 7]
```diff
@@ -49,19 +52,73 @@ private SelectClause CreateSelectClause(CodeElementsParser.Sql_selectClauseConte
...
+            }
+
             return null;
+        }
+
+        public DotStarSelection CreateDotStarSelection(CodeElementsParser.SelectionContext context)
+        {
+            if (context.dotStarSelection() != null)
+            {
+                if ((context.dotStarSelection().tableOrViewOrCorrelationName().Name != null))
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs:96]: Put `context.dotStarSelection().tableOrViewOrCorrelationName()` into a variable to avoid calling theses methods multiple times. Also when using the `as` operator you don't have to chekch for null before. A null value can be cast into any reference type so this can be simplified as:

```csharp
if (context.dotStarSelection()?.tableOrViewOrCorrelationName() != null)
{
    var tableOrViewOrCorrelationName = context.dotStarSelection().tableOrViewOrCorrelationName();
    Token name = tableOrViewOrCorrelationName.Name as Token;
    Token schemaName = tableOrViewOrCorrelationName.SchemaName as Token;
    Token dbms = tableOrViewOrCorrelationName.Name.DBMS as Token;
}
```

File: TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs [hunk 8]
```diff
@@ -10,5 +16,136 @@ public CodeElement CreateCommitStatement(CodeElementsParser.CommitStatementConte
...
+                        return (new QualifiedSymbolReference(qualifier, name));
+                    }
+                }
+                else
+                {
+                    return (name);
+                }
+            }
+
+            return (null);
```
Comment (fm-117) [inline TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs:148]: redundant parentheses

File: TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs [hunk 9]
```diff
@@ -10,5 +16,136 @@ public CodeElement CreateCommitStatement(CodeElementsParser.CommitStatementConte
...
+            return null;
+
+        }
+
+        private SymbolReference CreateSymbolReference(Token nameToken, Token qualifierToken, Token topLevelQualifierToken)
+        {
+            if (nameToken != null)
+            {
+                SymbolReference name = new SymbolReference(new AlphanumericValue(nameToken),
+                    SymbolType.SqlTableOrViewOrCorrelationName);
```
Comment (fm-117) [inline TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs:123]: I'm not sure this SymbolType is appropriate because it a _part_ of a table or view name. The `SqlTableOrViewOrCorrelationName`  suits for the whole name but not necessarily for its parts. So either we add multiple types like `SqlDbmsName`, `SqlSchemaName`, `SqlTableName`, `SqlColumnName`, or we use `SqlIdentifier` as a generic SymbolType. @smedilol, @mayanje what do you think ?

**Review by fm-117 — CHANGES_REQUESTED** [review]

File: TypeCobol.Test/Parser/CodeElements/EXEC.txt (hunk 4 above)
Comment (smedilol) [reply TypeCobol.Test/Parser/CodeElements/EXEC.txt:6]: This file is no longer in the PR, so it's ok.

File: TypeCobol.Test/Parser/Scanner/TestRealPrograms.cs [hunk 10]
```diff
@@ -1,5 +1,6 @@
 ﻿using System;
 using System.IO;
+using System.Linq;
```
Comment (fm-117) [inline TypeCobol.Test/Parser/Scanner/TestRealPrograms.cs:3]: Useless

File: TypeCobol.Test/TestCollection.cs (hunk 5 above)
Comment (fm-117) [reply TypeCobol.Test/TestCollection.cs:163]: Yes. @laabidiHend could you please revert this file to its original state ?

File: TypeCobol/AntlrGrammar/CobolWords.g4 (hunk 3 above)
Comment (fm-117) [inline TypeCobol/AntlrGrammar/CobolWords.g4:883]: Case is fine but indent is not, could you check that ?

File: TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs [hunk 11]
```diff
@@ -49,19 +57,86 @@ private SelectClause CreateSelectClause(CodeElementsParser.Sql_selectClauseConte
...
+
             if (context.star() != null)
             {
                 return new SelectClause(selectionModifier, new StarSelection());
             }
 
+            if (context.selections() != null)
+            {
+                var dotStarSelections = new List<DotStarSelection>();
+                int l = context.selections().ChildCount;
```
Comment (fm-117) [inline TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs:77]: not used

**Review by fm-117 — CHANGES_REQUESTED** [review]

File: TypeCobol.Test/Parser/Programs/Cobol85/ExecSql/ExecSqlWithSelectDotStar.rdz.cbl [hunk 12]
```diff
@@ -0,0 +1,11 @@
+       IDENTIFICATION DIVISION.
+       PROGRAM-ID. DVZZMFT3.
+       DATA DIVISION.
+       WORKING-STORAGE SECTION.
+         EXEC SQL 
+        SELECT ensi.hend.laabidi.* , khaled.*
```
Comment (fm-117) [inline TypeCobol.Test/Parser/Programs/Cobol85/ExecSql/ExecSqlWithSelectDotStar.rdz.cbl:6]: Please add a two-part qualified name here so and check that we correctly capture all parts in the correct tokens.

File: TypeCobol/AntlrGrammar/CobolWords.g4 (hunk 3 above)
Comment (fm-117) [reply TypeCobol/AntlrGrammar/CobolWords.g4:883]: By default Visual Studio replaces tabs with spaces so use Notepad to ensure the indent is correct.

File: TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs [hunk 13]
```diff
@@ -1,6 +1,9 @@
-﻿using TypeCobol.Compiler.AntlrUtils;
+﻿using System.Collections.Generic;
```
Comment (fm-117) [inline TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs:1]: `using System.Collections.Generic;` is redundant

File: TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs [hunk 14]
```diff
@@ -49,19 +56,77 @@ private SelectClause CreateSelectClause(CodeElementsParser.Sql_selectClauseConte
...
             }
+
             if (context.star() != null)
             {
                 return new SelectClause(selectionModifier, new StarSelection());
             }
 
+            if (context.selections() != null)
+            {
+                var dotStarSelections = context.selections().selection().Select(selection => CreateDotStarSelection(selection)).Where(selection => selection != null).ToList();
```
Comment (fm-117) [inline TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs:75]: You can simplify this a line a bit: `.Select(CreateDotStarSelection)`

File: TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs [hunk 15]
```diff
@@ -49,19 +56,77 @@ private SelectClause CreateSelectClause(CodeElementsParser.Sql_selectClauseConte
...
+            }
+            return null;
+        }
+
+        private DotStarSelection CreateDotStarSelection(CodeElementsParser.SelectionContext context)
+        {
+            if (context.dotStarSelection() != null)
+            {
+                var tableOrViewOrCorrelationName = context.dotStarSelection().tableOrViewOrCorrelationName();
+                if (tableOrViewOrCorrelationName.Name != null)
```
Comment (fm-117) [inline TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs:87]: I'm not so sure about this `if`. Event if Name is null, I think we should create a `DotStarSelection` instance (it would have a null `TableOrViewOrCorrelationName` property but that's ok) because it describes more precisely what's written. For example: `select tableName.*, .*` should lead to 2 `DotStarSelection` instance even though the second one is invalid.

**Review by fm-117 — CHANGES_REQUESTED** [review]

**Review by fm-117 — APPROVED** [review]

File: TypeCobol/AntlrGrammar/CobolCodeElements.g4 [hunk 16]
```diff
@@ -8265,8 +8265,11 @@ selectStatement: fullselect;
 fullselect: subselect;
 subselect: sql_selectClause;
 sql_selectClause: 
-SQL_SELECT (SQL_ALL|SQL_DISTINCT)? star;
-
+  SQL_SELECT (SQL_ALL|SQL_DISTINCT)? (star | selections) ;
+selections: selection (SQL_CommaSeparator selection)*;
+selection: dotStarSelection;
+dotStarSelection: tableOrViewOrCorrelationName  dot star; 
+tableOrViewOrCorrelationName : ((DBMS=UserDefinedWord dot)? (SchemaName=UserDefinedWord dot))? (Name=UserDefinedWord);
```
Comment (smedilol) [inline TypeCobol/AntlrGrammar/CobolCodeElements.g4:8272]: Can you add links to DB2 documentation in order to see the syntax rules defined by IBM ?
At least for `sql_selectClause` and `tableOrViewOrCorrelationName`.

**Review by smedilol — CHANGES_REQUESTED** [review]

File: TypeCobol/Compiler/Sql/Model/SelectClause.cs [hunk 17]
```diff
@@ -41,7 +42,13 @@ public class UnpackedRowSelection : Selection
 
     public class DotStarSelection : Selection
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol/Compiler/Sql/Model/SelectClause.cs:43]: Ok the intention is to select all columns from a given table or view so `TableOrViewAllColumnsSelection` ?

[the remark this answers is not in the document]
Comment (smedilol) [reply TypeCobol/Compiler/Sql/Model/SelectClause.cs:43]: Ok for `TableOrViewAllColumnsSelection`

File: TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs [hunk 18]
```diff
@@ -49,19 +55,76 @@ private SelectClause CreateSelectClause(CodeElementsParser.Sql_selectClauseConte
...
+
+        private TableOrViewAllColumnsSelection CreateDotStarSelection(CodeElementsParser.SelectionContext context)
+        {
+            if (context.dotStarSelection() != null)
+            {
+                var tableOrViewOrCorrelationName = context.dotStarSelection().tableOrViewOrCorrelationName();
+                Token name = tableOrViewOrCorrelationName.Name as Token;
+                Token schemaName = tableOrViewOrCorrelationName.SchemaName as Token;
+                Token dbms = tableOrViewOrCorrelationName.DBMS as Token;
+                SymbolReference fullName = CreateSymbolReference(name, schemaName, dbms);
```
Comment (mayanje) [inline TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs:89]: I think here it is better to create a class TableViewCorrelationName that ihnerits from QualifiedSymbolReference.
So that we can add getters on properties like DBMS, SchemaName, or Name.

Comment (mayanje) [reply TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs:89]: So I keep on thinking let's TableViewCorrelationName inherits from SqlObject rather than QualifiedSymbolReference.
I Suggest to have a class SqlStorageArea that ihnerits from StorageArea in this case TableViewCorrelationName can have an attribute SqlStorageArea built from QualifiedSymbolReference.
image

File: TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs [hunk 19]
```diff
@@ -78,7 +78,7 @@ private SelectClause CreateSelectClause(CodeElementsParser.Sql_selectClauseConte
             return null;
         }
 
-        private DotStarSelection CreateDotStarSelection(CodeElementsParser.SelectionContext context)
+        private TableOrViewAllColumnsSelection CreateDotStarSelection(CodeElementsParser.SelectionContext context)
```
Comment (fm-117) [inline TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs:81]: As this method will evolve to handle other type of selection, you should rename it `CreateSelection`. `CreateDotStarSelection` is not relevant anymore.

File: TypeCobol/Compiler/Sql/Model/SelectClause.cs [hunk 20]
```diff
@@ -40,12 +40,12 @@ public class UnpackedRowSelection : Selection
         public override SelectionType Type => SelectionType.UnpackedRow;
     }
 
-    public class DotStarSelection : Selection
+    public class TableOrViewAllColumnsSelection : Selection
     {
         public SymbolReference TableOrViewOrCorrelationName { get; }
         public override SelectionType Type => SelectionType.DotStar;
```
Comment (fm-117) [inline TypeCobol/Compiler/Sql/Model/SelectClause.cs:46]: Rename the enum also to `TableOrViewAllColumns`.

**Review by fm-117 — CHANGES_REQUESTED** [review]

File: TypeCobol/Compiler/Sql/Model/SelectClause.cs (hunk 20 above)
Comment (fm-117) [reply TypeCobol/Compiler/Sql/Model/SelectClause.cs:46]: The g4 grammar must be updated also to reflect the renaming. Do a global text search to locate the parts that need to be renamed.

File: TypeCobol/Compiler/CodeElements/Expressions/StorageArea.cs [hunk 21]
```diff
@@ -70,6 +72,15 @@ public override string ToString()
...
     }
 
+    public class SqlStorageArea : StorageArea
+    {
+        public SqlStorageArea(SymbolReference fullNameSymbolReference) : base()
+        {
+            this.SymbolReference = fullNameSymbolReference;
+        }
+        
+    }
```
Comment (fm-117) [inline TypeCobol/Compiler/CodeElements/Expressions/StorageArea.cs:82]: As discussed with @smedilol and @mayanje, we'll cover this new concept in a dedicated issue. Please revert the changes in this file.

File: TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs [hunk 22]
```diff
@@ -49,19 +55,85 @@ private SelectClause CreateSelectClause(CodeElementsParser.Sql_selectClauseConte
...
+                Token dbms = tableOrViewOrCorrelationName.DBMS as Token;
+                TableViewCorrelationName fullName = CreateSymbolReference(name, schemaName, dbms);
+                return new TableOrViewAllColumnsSelection(fullName);
+
+            }
+
             return null;
+        }
 
+        private TableViewCorrelationName CreateSymbolReference(Token nameToken, Token qualifierToken,
```
Comment (fm-117) [inline TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs:100]: In this PR we'll keep `TableViewCorrelationName` class but discard `SqlStorageArea` so to help factorize the code:
- revert to `private QualifiedSymbolReference CreateSymbolReference(Token nameToken, Token qualifierToken, Token topLevelQualifierToken)` so as to keep a generic method able to build qualified names
- wrap the result of this method into a `TableViewCorrelationName` in the `CreateSelection` method
  - the `TableViewCorrelationName` will now accept a QualifiedSymbolReference argument in its constructor instead of a `SqlStorageArea`

**Review by mayanje — APPROVED** [review]

File: TypeCobol/Compiler/Sql/Model/SqlObject.cs [hunk 23]
```diff
@@ -21,4 +22,18 @@ public bool AcceptVisitor(ISqlVisitor visitor)
...
+
+        public TableViewCorrelationName(SqlStorageArea tableViewOrCorrelation)
+        {
+            _tableViewOrCorrelation = tableViewOrCorrelation;
+        }
+        protected override bool VisitSqlObject(ISqlVisitor visitor)
+        {
+            return visitor.Visit(this);
+        }
+    }
```
Comment (fm-117) [inline TypeCobol/Compiler/Sql/Model/SqlObject.cs:38]: See my previous remark about this class. Also move it into its own file: add a 'Names.cs' file into '\TypeCobol\TypeCobol\Compiler\Sql\Model' and define the class in it. this file will be reused later for column name class.

**Review by fm-117 — CHANGES_REQUESTED** [review]

File: TypeCobol/Compiler/Sql/Model/SqlObject.cs [hunk 24]
```diff
@@ -1,5 +1,6 @@
 ﻿using System.Collections.Generic;
 using JetBrains.Annotations;
+using TypeCobol.Compiler.CodeElements;
```
Comment (fm-117) [inline TypeCobol/Compiler/Sql/Model/SqlObject.cs:3]: useless

File: TypeCobol/Compiler/Sql/Model/SqlObject.cs [hunk 25]
```diff
@@ -21,4 +22,6 @@ public bool AcceptVisitor(ISqlVisitor visitor)
 
         protected abstract bool VisitSqlObject([NotNull] ISqlVisitor visitor);
     }
+
+
```
Comment (fm-117) [inline TypeCobol/Compiler/Sql/Model/SqlObject.cs:26]: redundant change, please revert

File: TypeCobol/Compiler/CodeElements/Expressions/StorageArea.cs [hunk 26]
```diff
@@ -70,6 +70,7 @@ public override string ToString()
         }
     }
 
+
```
Comment (fm-117) [inline TypeCobol/Compiler/CodeElements/Expressions/StorageArea.cs:73]: redundant change, to revert

File: TypeCobol/Compiler/Sql/Model/TableViewCorrelationName .cs [hunk 27]
```diff
@@ -0,0 +1,23 @@
...
+        public TableViewCorrelationName(SymbolReference tableViewOrCorrelation)
+        {
+            _tableViewOrCorrelation = tableViewOrCorrelation;
+        }
+        protected override bool VisitSqlObject(ISqlVisitor visitor)
+        {
+            return visitor.Visit(this);
+        }
+    }
+}
```
Comment (fm-117) [inline TypeCobol/Compiler/Sql/Model/TableViewCorrelationName .cs:23]: Could you please rename the file to `Names.cs`, it will contain future sql name definitions.
Also remove useless using directives (everything except using TypeCobol.Compiler.CodeElements).

File: TypeCobol/Compiler/Sql/Model/TableViewCorrelationName .cs [hunk 28]
```diff
@@ -0,0 +1,23 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using TypeCobol.Compiler.CodeElements;
+
+namespace TypeCobol.Compiler.Sql.Model
+{
+    public class TableViewCorrelationName : SqlObject
+    {
+        private SymbolReference _tableViewOrCorrelation;
```
Comment (fm-117) [inline TypeCobol/Compiler/Sql/Model/TableViewCorrelationName .cs:12]: readonly field

**Review by fm-117 — CHANGES_REQUESTED** [review]

File: TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs [hunk 29]
```diff
@@ -49,19 +55,85 @@ private SelectClause CreateSelectClause(CodeElementsParser.Sql_selectClauseConte
...
+
+            return null;
+        }
+
+        private TableOrViewAllColumnsSelection CreateSelection(CodeElementsParser.SelectionContext context)
+        {
+            if (context.tableOrViewAllColumnsSelection() != null)
+            {
+                var tableOrViewOrCorrelationName =
+                    context.tableOrViewAllColumnsSelection().tableOrViewOrCorrelationName();
```
Comment (fm-117) [inline TypeCobol/Compiler/Sql/CodeElements/SqlCodeElementBuilder.cs:88]: Could you test what happens on a syntax like `SELECT .*` ? Does ANTLR produce an error and we don't even enter this method or `context.tableOrViewAllColumnsSelection().tableOrViewOrCorrelationName()`is null and then we have an exception ?

File: TypeCobol/Compiler/Sql/Model/SelectClause.cs [hunk 30]
```diff
@@ -39,9 +40,15 @@ public class UnpackedRowSelection : Selection
         public override SelectionType Type => SelectionType.UnpackedRow;
     }
 
-    public class DotStarSelection : Selection
+    public class TableOrViewAllColumnsSelection : Selection
     {
-        public override SelectionType Type => SelectionType.DotStar;
+        public TableViewCorrelationName TableOrViewOrCorrelationName { get; }
```
Comment (fm-117) [inline TypeCobol/Compiler/Sql/Model/SelectClause.cs:45]: If `TableViewCorrelationName` is an SqlObject, Selection and all its derived classes must inherit from SqlObject too. Otherwise the visitor pattern won't work

**Review by fm-117 — CHANGES_REQUESTED** [review]

**Review by fm-117 — APPROVED** [review]

**Review by smedilol — APPROVED** [review]
